### PR TITLE
kloud/provider: Add SoftLayer Support

### PR DESCRIPTION
**Pull request by acbodine**
Adds SoftLayer support as a kloud provider, and necessary frontend changes.

## Description
Following the example here [https://github.com/koding/kloud-provider-example](https://github.com/koding/kloud-provider-example) and the PRs for [DigitalOcean](https://github.com/koding/koding/pull/9187) and [Google](https://github.com/koding/koding/pull/9185) support, this PR include the necessary changes to enable SoftLayer provider support for kloud.

Part of this PR includes updates to the vendor Terraform source in this repository, and essentially consists of getting the SoftLayer provider directory from Terraform's master branch and dropping it in place here. The current SoftLayer support that Terraform has is basically provisioning virtual guests (or vms) and managing ssh keys.

This PR also includes most (maybe all?) of the UI/frontend changes necessary for a user to use Koding with SoftLayer.

## Motivation and Context
Implement a solution to #9073.

## How Has This Been Tested?
- Unit tests to validate basic functionality for `stack` `machine` and `schema` definitions
- Integration tests that rely on environment variables to actually call SoftLayer apis where necessary in validating `stack` and `machine` definitions
- I found mentions on Slack channel about the `downtowork` koding development deployment, so I created a development environment [http://downtowork-acbodine.koding.team:8090](http://downtowork-acbodine.koding.team:8090). In order to test my changes with the whole system I simply pulled my changes into this enviroment and rebuilt/re-ran the necessary Docker containers. You can create a team there, or use the one I have been developing with [http://dragons.downtowork-acbodine.koding.team:8090](http://dragons.downtowork-acbodine.koding.team:8090) and you can leverage the SoftLayer stack I have in that team `Voltorb SoftLayer Stack`
- The frontend/user testing that I did has been manual so far in order to verify that I could
  1. Verify Credentials user enters
  2. Save SoftLayer stack and generate Bootstrap resources, verified in SoftLayer UI
  3. Initialize SoftLayer virtual guest, and verified connectivity back to Koding client/UI
  4. Tested Koding UI displaying accurate virtual guest state info, as well as stop/start actions

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: client/home/lib/virtualmachines/components/machineslist/machinedetails.coffee
Comment (gokmen) [inline client/home/lib/virtualmachines/components/machineslist/machinedetails.coffee]: 👍

**Review by gokmen — APPROVED** [review]
LGTM on node/client side 👍

File: go/src/koding/kites/kloud/provider/softlayer/stack_test.go
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/softlayer/stack_test.go]: Could you add a test that verifies the template before and after applying?

Like [this](https://github.com/koding/koding/blob/c38a33244c85d6721ef8796d051e2f8b9004e496/go/src/koding/kites/kloud/provider/marathon/stack_test.go#L28) or [this](https://github.com/koding/koding/blob/c38a33244c85d6721ef8796d051e2f8b9004e496/go/src/koding/kites/kloud/provider/azure/stack_test.go#L89) one.

File: go/src/koding/kites/kloud/provider/softlayer/stack.go
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/softlayer/stack.go]: Not needed. Shadowing is there to prevent leaking delagated credentials.

If you have an admin and a user within your team, and user should now know Softlayer credentials used to create resources, then admin creates a credential and shares it with the user. Since the template is augemented server-side, user never sees the real credentials used.

The shadowing is used to prevent malicious user to extract credentials from template, e.g. by echoing them to a file during vm creating.

Since user has no access currently to bootstrap templates, no need for shadowing.

File: go/src/koding/kites/kloud/provider/softlayer/stack.go
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/softlayer/stack.go]: In case of Softlayer probably it's not that important. The softlayer-go library misuses Basic Auth, instead of adding a header to a HTTP request it adds auth info to the hostname:

https://github.com/maximilien/softlayer-go/blob/master/client/http_client.go#L69

So if request times out, go's `net/http` library is logging the whole hostname with `*net.OpError` errors. So, until the lib is fixed, the credentials will leak. Just FYI.

File: go/src/koding/kites/kloud/provider/softlayer/stack.go
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/softlayer/stack.go]: Is `1234567` always a valid object in Softlayer?

File: go/src/koding/kites/kloud/provider/softlayer/stack.go
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/softlayer/stack.go]: Bootstrap templates are fully controlled by kloud currently (server-side), user can't edit it now.

We're working actively to make klient support more distros, just FYI.

**Review by rjeczalik — APPROVED** [review]
Looks good, just it would be ideal to have blanket test for ApplyTemplate as described above. But no crucial.

File: go/src/koding/kites/kloud/provider/softlayer/stack.go
Comment (acbodine) [inline go/src/koding/kites/kloud/provider/softlayer/stack.go]: Thanks for clearing that up, I'll remove the `NOTE`

File: go/src/koding/kites/kloud/provider/softlayer/stack.go
Comment (acbodine) [inline go/src/koding/kites/kloud/provider/softlayer/stack.go]: I sort of know the guy who maintains the `softlayer-go` repo, I'll see if I can reach out to him to get this fixed. I'll leave the shadowing in place for now.

File: go/src/koding/kites/kloud/provider/softlayer/stack.go
Comment (acbodine) [inline go/src/koding/kites/kloud/provider/softlayer/stack.go]: Correct

File: go/src/koding/kites/kloud/provider/softlayer/softlayer_test.go
Comment (acbodine) [inline go/src/koding/kites/kloud/provider/softlayer/softlayer_test.go]: Good catch, deleting this.

File: go/src/koding/kites/kloud/provider/softlayer/stack_test.go
Comment (acbodine) [inline go/src/koding/kites/kloud/provider/softlayer/stack_test.go]: I'll get this test added in asap.

Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/softlayer/stack_test.go]: You can copy `stripNondeterministicResources` and `equal` to your package and adjust them to your needs.

If you created `stack/provider/providertest` package and move them there to have only one copy of them, it would be really awesome.

Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/softlayer/stack_test.go]: (in the meantime I've added `object.ReplaceFunc` which does more-or-less what `stripNondeterministicResources` does but in more generic manner - it may be useful here)

**Comment by acbodine** [comment]
@rjeczalik I added the blanket test for ApplyTemplate and refactored the template verification logic you pointed me at to `stack/provider/providertest` package. Let me know if anything else needs to be done.

**Comment by gokmen** [comment]
@acbodine please ignore DO-NOT-MERGE label it's related with deployment process, we'll merge this in once it's ready after the next deploy, fyi.

File: go/src/koding/kites/kloud/provider/softlayer/stack.go
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/softlayer/stack.go]: Please include team name also here:

```
fmt.Sprintf("koding-%s-%s", s.Arg(*stack.BootstrapRequest).GroupName, c.Identifier)
```

File: go/src/koding/kites/kloud/provider/softlayer/stack_test.go
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/softlayer/stack_test.go]: The file name should be `single-guest.json.golden`.

File: go/src/koding/kites/kloud/stack/provider/providertest/apply.go
Comment (rjeczalik) [inline go/src/koding/kites/kloud/stack/provider/providertest/apply.go]: Please gofmt this file.

**Comment by rjeczalik** [comment]
@acbodine Looks good, just minor comments.

**Comment by acbodine** [comment]
@rjeczalik Those should be all set now.

**Comment by rjeczalik** [comment]
@acbodine Thanks, this is really good one! 💥

**Comment by rjeczalik** [comment]
@acbodine Thanks, this is really a good one! 💥

**Review by rjeczalik — APPROVED** [review]

**Comment by acbodine** [comment]
@gokmen I resolved the merge conflicts

**Comment by gokmen** [comment]
thanks @acbodine we're waiting for softlayer test accounts, after tests @rjeczalik will merge this one 👍

**Comment by rjeczalik** [comment]
Credential verification would need to be fixed, right now it's a false positive and during bootstrap we are getting `Access denied` error during creating SSH keys.

We can merge this and iterate on fixes in separate PRs.

/cc @gokmen @cihangir

**Comment by rjeczalik** [comment]
@acbodine Thanks for Softlayer, really nice work!

**Pull request merged**
